datastore/cache.go

```diff
@@ -0,0 +1,151 @@
+/*
+AUTHORS
+  Alan Noble <alan@ausocean.org>
+
+LICENSE
+  Copyright (c) 2023, The OpenFish Contributors.
+
+  Redistribution and use in source and binary forms, with or without
+  modification, are permitted provided that the following conditions are met:
+
+  1. Redistributions of source code must retain the above copyright notice, this
+     list of conditions and the following disclaimer.
+
+  2. Redistributions in binary form must reproduce the above copyright notice,
+     this list of conditions and the following disclaimer in the documentation
+     and/or other materials provided with the distribution.
+
+  3. Neither the name of The Australian Ocean Lab Ltd. ("AusOcean")
+     nor the names of its contributors may be used to endorse or promote
+     products derived from this software without specific prior written permission.
+
+  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+  DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+  FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+  DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+  SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+  OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+  OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*/
+
+package datastore
+
+import (
+	"errors"
+	"sync"
+)
+
+// Type Cache defines the (optional) caching interface used by Entity.
+type Cache interface {
+	Set(key *Key, value Entity)   // Set adds or updates a value to the cache.
+	Get(key *Key) (Entity, error) // Get retrieves a value from the cache, or returns ErrcacheMiss.
+	Delete(key *Key)              // Delete removes a value from the cache.
+	Reset()                       // Reset resets (clears) the cache.
+}
+
+// Type genericCache represents a generic cache for holding datastore entities.
+// The key K is either an int64 or a string.
+type genericCache[K comparable] struct {
```

Comment:
I'm thinking now that this should go in one of two ways and it's currently inbetween. 

(first way) If it's going to be unexported, and be used only by this file, then you should either use a constraint, like 
```
type KeyConstraint interface {
    int64 | string
}
```
instead of comparable, or do away with the generic type all together and defaulting to keying using Key. I actually don't see a need for two separate caches as ID cache and Name cache; you can use the Key type given that it's a struct and is comparable.

(second way) make this cache more available for use outside of this package, i.e. make exported, and make both key and value types generic i.e. type GenericCache[K comparable, V any] struct.  But, even in this case, you can still have a single cache type keying using Key.

Reply:
I think I prefer the first approach for its simplicity. Google's datastore.Key is not currently comparable because it does not implement Less (only Equals). Are you suggesting what we implement our own Key?

Reply:
I mean comparable in the sense that you can use it as a map key type. 

In the following example, I've got the actual google datastore.Key struct and used it as a key type 
https://go.dev/play/p/EPChFTZ_G1X

Reply:
I think comparable in go just means that you can do == , which you can do with structs.

Reply:
> I mean comparable in the sense that you can use it as a map key type.
> 
> In the following example, I've got the actual google datastore.Key struct and used it as a key type https://go.dev/play/p/EPChFTZ_G1X

Nice!